src/Umbraco.Infrastructure/Persistence/Repositories/Implement/TrackedReferencesRepository.cs

```diff
@@ -0,0 +1,194 @@
+using System.Collections.Generic;
+using System.Linq;
+using NPoco;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Persistence;
+using Umbraco.Cms.Core.Persistence.Repositories;
+using Umbraco.Cms.Core.Scoping;
+using Umbraco.Cms.Infrastructure.Persistence.Dtos;
+using Umbraco.Extensions;
+
+namespace Umbraco.Cms.Infrastructure.Persistence.Repositories.Implement
+{
+    internal class TrackedReferencesRepository : ITrackedReferencesRepository
+    {
+        private readonly IScopeAccessor _scopeAccessor;
+
+        public TrackedReferencesRepository(IScopeAccessor scopeAccessor)
+        {
+            _scopeAccessor = scopeAccessor;
+        }
+
+        public IEnumerable<RelationItem> GetPagedItemsWithRelations(int[] ids, long pageIndex, int pageSize,
+            bool filterMustBeIsDependency, out long totalRecords)
+        {
+            var sql = _scopeAccessor.AmbientScope.Database.SqlContext.Sql().Select(
+                    "[pn].[id] as nodeId",
+                    "[pn].[uniqueId] as nodeKey",
+                    "[pn].[text] as nodeName",
+                    "[pn].[nodeObjectType] as nodeObjectType",
+                    "[ct].[icon] as contentTypeIcon",
+                    "[ct].[alias] as contentTypeAlias",
+                    "[ctn].[text] as contentTypeName",
+                    "[umbracoRelationType].[alias] as relationTypeAlias",
+                    "[umbracoRelationType].[name] as relationTypeName",
+                    "[umbracoRelationType].[isDependency] as relationTypeIsDependency",
+                    "[umbracoRelationType].[dual] as relationTypeIsBidirectional")
+                .From<RelationDto>("r")
+                .InnerJoin<RelationTypeDto>("umbracoRelationType").On<RelationDto, RelationTypeDto>((left, right) => left.RelationType == right.Id, aliasLeft: "r", aliasRight:"umbracoRelationType")
+                .InnerJoin<NodeDto>("cn").On<RelationDto, NodeDto, RelationTypeDto>((r, cn, rt) => (!rt.Dual && r.ParentId == cn.NodeId) || (rt.Dual && (r.ChildId == cn.NodeId || r.ParentId == cn.NodeId )), aliasLeft: "r", aliasRight:"cn", aliasOther: "umbracoRelationType" )
+                .InnerJoin<NodeDto>("pn").On<RelationDto, NodeDto, NodeDto>((r, pn, cn) => (pn.NodeId == r.ChildId && cn.NodeId == r.ParentId) || (pn.NodeId == r.ParentId && cn.NodeId == r.ChildId), aliasLeft: "r", aliasRight:"pn", aliasOther:"cn" )
+                .LeftJoin<ContentDto>("c").On<NodeDto, ContentDto>((left, right) => left.NodeId == right.NodeId, aliasLeft:"pn", aliasRight:"c")
+                .LeftJoin<ContentTypeDto>("ct").On<ContentDto, ContentTypeDto>((left, right) => left.ContentTypeId == right.NodeId, aliasLeft:"c",  aliasRight:"ct")
+                .LeftJoin<NodeDto>("ctn").On<ContentTypeDto, NodeDto>((left, right) => left.NodeId == right.NodeId, aliasLeft:"ct",  aliasRight:"ctn");
+
+            if (ids.Any())
+            {
+                sql = sql.Where<NodeDto>(x => ids.Contains(x.NodeId), "pn");
```

Comment:
Perhaps this could use sql.WhereIn

Reply:
Checking the resulting SQL statement actually converts the above to <code><b>WHERE</b> ([pn].[id] <b>IN</b> (. . .)</code>. Is there any specific reason why you would prefer to use `sql.WhereIn`?

Reply:
No strong preference, code may be easier to rewrite per database type when using the available extension methods.